// powwow-server-common/logger.js
'use strict';

// Dependencies:
var fs = require('fs');
var path = require('path');
var bunyan = require('bunyan');
var moment = require('moment');
var serverConfig = require('./serverConfig');
var launchparams = require('./launchparams');
// package details:
var pkg = require.main.require('./package');

// Prepare output streams:
var streams = [];
streams.push({
    stream: process.stdout,
    level: serverConfig.logLevel
});

/**
 * @launchparam persistlogs
 */
if(launchparams.persistlogs){
    // Save logs for accessing them later.
    var logDir = path.join(process.cwd(), "logs");
    if (!fs.existsSync(logDir)) fs.mkdirSync(logDir);
    var newLogFilename = "dev-" + moment().format("YYYY-MM-DD-HH.mm.ss-SSS") + "-" + process.pid;
    // Create log directory if it doesn't exist
    streams.push({
        path: path.join(logDir, newLogFilename),
        level: 'debug'
    });
}

// initialize logger with configured streams:
var logger = bunyan.createLogger({
    name: pkg.name,
    streams: streams,
    serializers: bunyan.stdSerializers
});

logger.DEBUG = bunyan.DEBUG;
logger.INFO = bunyan.INFO;
logger.ERROR = bunyan.ERROR;
logger.FATAL = bunyan.FATAL;

logger.info('Starting ' + pkg.name + ', version ' + pkg.version);

module.exports = logger;
