// powwow-server-common/phantomBrowser.js
'use strict';

/**
 * Dependencies.
 */

var phantom = require('./bridge/node-phantom-simple');
var logger = require('./logger');
var pageService = require('./pageService');
var config = require.main.require('./config');
var http = require('http');
var fs = require('fs');
var launchparams = require('./launchparams');

var phantomBrowser;
var mainPage = null;
var tempLocalStorageFolder = null;

exports.getBrowser = function () {
    return phantomBrowser;
};

exports.getMainPage = function () {
    return mainPage;
};

exports.setMainPage = function (page) {
    mainPage = page;
};

exports.createBrowserInstanceAndNewPage = function (proxyInfo, callback) {
    var parameters = {
        'ignore-ssl-errors': 'true',
        'web-security': 'false',
        'load-images': 'true' // This causes a memory leak when there are popups, so not allowing this to be set for now.
    };

    if(config.cache && config.cache.enabled) {
        parameters['disk-cache'] = 'true';
        if(config.cache.size) {
            parameters['max-disk-cache-size'] = '' + config.cache.size;
        }
        if(config.cache.path) {
            parameters['disk-cache-path'] = config.cache.path;
        }
    }

    // Set up remote debugging.
    if (config.remoteDebug && config.remoteDebug.enabled) {
        parameters['debug'] = config.remoteDebug.phantomRequestResponse ? 'true' : 'false';
        parameters['remote-debugger-port'] = '' + config.remoteDebug.port; // This should be a string.
        parameters['remote-debugger-autorun'] = 'yes';
    }

    // Isolate localStorage for this instance of Phantom JS.
    if(!config.localStorageIsolationOff) {
        var basePath = config.localStoragePath ? (config.localStoragePath + "/") : ".localStorage/";

        if (launchparams.sessionID) {
            tempLocalStorageFolder = basePath + launchparams.sessionID;
        } else {
            var randomDir = Math.floor((Math.random()*1000000)) + "";
            tempLocalStorageFolder = basePath + randomDir;
        }
        logger.info("Setting local storage folder:", tempLocalStorageFolder);
        parameters['local-storage-path'] = tempLocalStorageFolder;
    }

    var createOptions = { parameters: parameters };
    // Set up optional passed in environment variables for Phantom JS.
    if (config.env) {
        createOptions.env = config.env;
    }

    if(config.userAgentReplacement) {
        createOptions.bridgeparameter = config.userAgentReplacement;
    }

    // Create browser instance:
    phantom.create(createOptions, function (err, browser) {
        if (err) {
            logger.info("Error creating phantom instance", err);
            return;
        }

        phantomBrowser = browser;

        logger.info("Config file:", JSON.stringify(config));

        if (proxyInfo.host) {
            logger.info("Proxy:", JSON.stringify(proxyInfo));
            phantomBrowser.setProxy(proxyInfo.host, proxyInfo.port, proxyInfo.proxyType, proxyInfo.user, proxyInfo.password);
        }

        // If cookies are provided during launch, add the cookies 
        if (launchparams.hasOwnProperty('cookies')) {
            for (var cookie of launchparams.cookies) {
                phantomBrowser.addCookie(cookie);
            }
        }

        // Now create and load the main page.
        pageService.createPhantomPage(browser, function (page) {
            callback(page);
            logger.info('Start URL is', config.startURL);
            page.open(config.startURL, function (status) {
                //logger.info('Web-page status is', status);
                if (status === 'success') {
                    logger.info('Web-page is ready.');
                }
            });
        });
    });
};

exports.close = function () {
    logger.info("Closing browser");
    if (phantomBrowser) {
        phantomBrowser.exit();
    }
    if(tempLocalStorageFolder) {
        try {
            // Clearing local storage folder.
            logger.info("Clearing local storage folder:", tempLocalStorageFolder);
            deleteFolderRecursive(config.localStoragePath);
        } catch(e) {}
    }
    process.exit();
};

exports.clearCookies = function (callback) {
    phantomBrowser.clearCookies();
};

function deleteFolderRecursive(path) {
  if( fs.existsSync(path) ) {
    fs.readdirSync(path).forEach(function(file,index){
      var curPath = path + "/" + file;
      if(fs.lstatSync(curPath).isDirectory()) { // recurse
        deleteFolderRecursive(curPath);
      } else { // delete file
        fs.unlinkSync(curPath);
      }
    });
    fs.rmdirSync(path);
  }
};
