// powwow-server-common/server.js
'use strict';

var ASQ = require('asynquence');

var api =  require.main.require('./api');

var logger = require('./logger');
var stateManager = require('./stateManager');
var errors = require('./errors');
var client = require('./client');
var Page = require('./page');
var Promise = require('promise');
var util = require('./util');

require('./asynquence-abort');

exports.onconnection = stateManager.onconnection;
exports.onconnectionclose = stateManager.onconnectionclose;

exports.onrequest = function(connection, request) {
    util.logRequest(request);

    if(request.method == ".initialize") {
        util.initialize(request.params);
        return;
    }

    if(request.method == ".checkstate") {
        util.checkstate(request.params);
        return;
    }

    var context = {
        connection : connection,
        request : request
    };
    var seq;

    seq = ASQ(context).then(stateManager.prepareContext).then(function(done, context) {
        var page = Page(context);
        dispatchMessage(page, context);
        done();
        });
};

var dispatchMessage = function(page, context) {
    var namespace = context.request.method.split('.')[0];
    var action = context.request.method.split('.')[1];
    var obj = (namespace.length == 0) ? util : api[namespace];

    if ( typeof obj === 'object' && typeof obj[action] === 'function') {
        try {
            stateManager.startRequest(page);
            obj[action].call(obj, page, context.request.params);
            // Not calling endRequest here on purpose.  End request happens after
            // the api call is fully done
        } catch (e) {
            client.notification('error')({
                error : e.message
            });
            logger.error(e.message);
        }

    } else {
        var msg = 'The method does not exist: ' + namespace + "." + action;
        client.notification('error')({
            error : msg
        });
        logger.error(msg);
    }
};

var sendResponse = function(context, result) {
    // no response to a notification
    if (context.request && context.request.type == 'notification') {
        logger.debug({
            result : result
        }, "Unsent response to notification");
        return;
    }

    result.jsonrpc = '2.0';

    if (context.request) {
        result.id = context.request.id;
    }

    logger.debug({
        response : result
    }, "Send response");
    context.connection.send(JSON.stringify(result));
};
